cpp/src/arrow/scalar.cc

```diff
@@ -241,6 +194,32 @@ Status CastImpl(const BooleanScalar& from, NumericScalar<T>* to) {
   return Status::OK();
 }
 
+// numeric to temporal
+template <typename From, typename To>
+typename std::enable_if<std::is_base_of<TemporalType, To>::value &&
+                            !std::is_same<DayTimeIntervalType, To>::value,
+                        Status>::type
+CastImpl(const NumericScalar<From>& from, TemporalScalar<To>* to) {
+  to->value = static_cast<typename To::c_type>(from.value);
+  return Status::OK();
+}
+
+// temporal to numeric
+template <typename From, typename To>
+typename std::enable_if<std::is_base_of<TemporalType, From>::value &&
+                            !std::is_same<DayTimeIntervalType, From>::value,
+                        Status>::type
+CastImpl(const TemporalScalar<From>& from, NumericScalar<To>* to) {
+  to->value = static_cast<typename To::c_type>(from.value);
+  return Status::OK();
+}
+
+// timestamp to timestamp
+Status CastImpl(const TimestampScalar& from, TimestampScalar* to) {
+  to->value = from.value;
```

Comment:
Should this accomodate potential differences in units? For example, if `from` is microseconds while `to` is nanoseconds, then we should have 
```suggestion
  to->value = from.value * 1000;
```